### PR TITLE
ci: Add markdownlint check

### PR DESCRIPTION
**Pull request by vadorovsky**
Let the machines nitpick the line lengths and whitespaces!

- - -
This change is [Reviewable](https://reviewable.io/reviews/aya-rs/book/165)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -24,6 +24,12 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      - name: Markdown lint
```
Comment (tamird) [inline .github/workflows/ci.yml:27]: should be
```
      - uses: DavidAnson/markdownlint-cli2-action@v17
        with:
          globs: 'docs/**/*.md'
```

the one you're using here is archived and not maintained

**Review by tamird — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (vadorovsky) [reply .github/workflows/ci.yml]: Done

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: all files reviewed, 13 unresolved discussions (waiting on @vadorovsky)

___
*[`README.md` line 10 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSqWpANRNVfocXq0Y:-O8hSqWpANRNVfocXq0Z:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/README.md#L10)):*
> ```Markdown
> 
> [build-badge]: https://img.shields.io/github/actions/workflow/status/aya-rs/book/ci.yml
> [build-url]: https://github.com/aya-rs/book/actions/workflows/ci.yml
> ```

relative?
___
*[`docs/index.md` line 64 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSx3FBCNxzd42jo1v:-O8hSx3FBCNxzd42jo1w:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/index.md#L64)):*
> ```Markdown
> [pulsar]: https://pulsar.sh/
> [pulsar-github]: https://github.com/Exein-io/pulsar
> [kubernetes-logo]: https://github.com/aya-rs/book/assets/5332524/abde6552-10ed-4c52-9717-732d1ec7ea6c
> ```

relative?
___
*[`docs/book/aya/aya-tool.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hStZC6nHHNcWgk_7k:-O8hStZC6nHHNcWgk_7l:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/aya-tool.md#L6)):*
> ```Markdown
> 
>     Full code for the example in this chapter is available
>     [here](https://github.com/aya-rs/book/tree/main/examples/aya-tool)
> ```

relative?
___
*[`docs/book/aya/crosscompile.md` line 12 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSZQGBtYmbI4XO9LF:-O8hSZQGBtYmbI4XO9LG:b-o3npi6) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/crosscompile.md#L12)):*
> ```Markdown
> 
> 1. Install `rustup` following the instructions on <https://rustup.rs/>
> 2. Install the stable and nightly rust toolchains:
> ```

do we need to number these manually? why?
___
*[`docs/book/aya/lifecycle.md` line 30 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSciKEt94Ce1M2FI8:-O8hSciKEt94Ce1M2FI9:b-huvmot) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/lifecycle.md#L30)):*
> ```Markdown
> 1. When you call `load` or `load_file`, all the maps referenced by the eBPF
>    code are created and stored inside the returned Bpf instance.
> 2. Similarly when you load a program to the kernel, it's stored inside the `Bpf`
> ```

why manually number these?
___
*[`docs/book/programs/classifiers.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSkPD6FiGP3JVituv:-O8hSkPD6FiGP3JVituw:b-jovpgd) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/classifiers.md#L6)):*
> ```Markdown
> 
>     Full code for the example in this chapter is available
>     [here](https://github.com/aya-rs/book/tree/main/examples/tc-egress)
> ```

can this be relative?
___
*[`docs/book/programs/lsm.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSmEG5koGQjmVmwt0:-O8hSmEG5koGQjmVmwt1:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/lsm.md#L6)):*
> ```Markdown
> 
>     Full code for the example in this chapter is available
>     [here](https://github.com/aya-rs/book/tree/main/examples/lsm-nice)
> ```

relative?
___
*[`docs/book/programs/probes.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSn4rDDCfrsH0cBjb:-O8hSn4rDDCfrsH0cBjc:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/probes.md#L6)):*
> ```Markdown
> 
>     Full code for the example in this chapter is available
>     [here](https://github.com/aya-rs/book/tree/main/examples/kprobetcp).
> ```

relative?
___
*[`docs/book/programs/xdp.md` line 5 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSnn94xQwZnj5iiYO:-O8hSnn94xQwZnj5iiYP:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/xdp.md#L5)):*
> ```Markdown
> !!! example "Source Code"
> 
>     Full code for the example in this chapter is available [here](https://github.com/aya-rs/book/tree/main/examples/xdp-drop)
> ```

relative?
___
*[`docs/book/start/dropping-packets.md` line 68 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hTxNk4CvHJDhTMwPJ:-O8hTxNlEJHNTL_xVoPL:bxma2sq) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/dropping-packets.md#L68)):*
> ```Markdown
> 
> 1. Get a reference to the map
> 2. Create an IPv4Addr
> ```

let's avoid manually numbering?
___
*[`docs/book/start/hello-xdp.md` line 5 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSvHE7GYvpiN5vqZf:-O8hSvHE7GYvpiN5vqZg:b-emtlki) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/hello-xdp.md#L5)):*
> ```Markdown
> !!! example "Source Code"
> 
>     Full code for the example in this chapter is available [here](https://github.com/aya-rs/book/tree/main/examples/xdp-hello)
> ```

relative?
___
*[`docs/book/start/hello-xdp.md` line 131 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hU2Go8tTPxWxkHrtr:-O8hU2Go8tTPxWxkHrts:bxma2sq) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/hello-xdp.md#L131)):*
> ```Markdown
>    [Ctrl-C handler][ctrl-c-handler]. It will come in useful later as we expand
>    the functionality of the initial program:
> 2. Here we declare our CLI flags. Just `--iface` for now for passing the
> ```

let's avoid manually numbering?
___
*[`examples/kprobetcp/README.md` line 10 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hU7FI8gpGbsR3fc3v:-O8hU7FJ1FdNdEKMqmyL:bxma2sq) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/examples/kprobetcp/README.md#L10)):*
> ```Markdown
> 
> 1. Install a rust stable toolchain: `rustup install stable`
> 2. Install a rust nightly toolchain: `rustup install nightly`
> ```

let's avoid manually numbering?

**Review by vadorovsky — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: all files reviewed, 13 unresolved discussions (waiting on @tamird)

___
*[`README.md` line 10 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSqWpANRNVfocXq0Y:-O8kWEha4Fi9QkareJLb:b-85oqdn) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/README.md#L10)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

Doesn't work. I tried and relative links are adding `/blob` to the URL, thinking that you're just trying to link to the file in the repo.
___
*[`docs/index.md` line 64 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSx3FBCNxzd42jo1v:-O8kRZaRAtR2fF79DS5j:b-tb60dt) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/index.md#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

I can't make this exact link relative in the book (see the comment below).

What I could do if you really want to have this link relative, would be including the Kubernetes logo in `docs/assets` (in a git repo, not as a Github asset, which I don't even know how it's being created :roll_eyes: ), so they end up in the mkdocs build. But I'm not sure if it's the right thing to do. Kubernetes logo is not our asset.

I would rather just link to https://raw.githubusercontent.com/cncf/artwork/refs/heads/main/projects/kubernetes/horizontal/color/kubernetes-horizontal-color.svg if you don't mind.
___
*[`docs/book/aya/aya-tool.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hStZC6nHHNcWgk_7k:-O8kS1jNDhnSNTYIU3dd:bzaagnk) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/aya-tool.md#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

Relative link won't work in the book. Book is totally independent from Github repo, you can compile it and host anywhere - like we do on aya-rs.dev. But you'll encounter issues also when doing a local mkdocs build. Github isn't relative to aya-rs.dev.
___
*[`docs/book/aya/crosscompile.md` line 12 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSZQGBtYmbI4XO9LF:-O8kPZoRBoddfUqpL_2W:bxl8xa3) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/crosscompile.md#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

do we need to number these manually? why?
</blockquote></details>

I swear that having all points as `1.` was triggering a markdownlint error when I started this PR, but I can't reproduce it anymore. Perhaps they relaxed that requirement in markdownlint-cli 0.42.0.

Also, in case of this file, there are points which are followed by code blocks. These code blocks are surrounded by blank lines, which breaks auto numbering. When blank lines are removed, auto numbering works, but `MD031/blanks-around-fences` gets triggered.

LMK what's your opinion about this, but I'm leaning towards removing the blank lines and disabling `MD031/blanks-around-fences` check in this part.
___
*[`docs/book/aya/lifecycle.md` line 30 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSciKEt94Ce1M2FI8:-O8kVpqC7Vj2-V8a351p:b-896fix) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/lifecycle.md#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why manually number these?
</blockquote></details>

Done.
___
*[`docs/book/programs/classifiers.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSkPD6FiGP3JVituv:-O8kViHx6I5XTaN4MUFr:bbfrbk4) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/classifiers.md#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can this be relative?
</blockquote></details>

nope, explained above
___
*[`docs/book/programs/lsm.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSmEG5koGQjmVmwt0:-O8kVkvL294WpopsnYzy:bbfrbk4) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/lsm.md#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

nope, explained above
___
*[`docs/book/programs/probes.md` line 6 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSn4rDDCfrsH0cBjb:-O8kVlK46_K2O67_31bB:bbfrbk4) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/probes.md#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

nope, explained above
___
*[`docs/book/programs/xdp.md` line 5 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSnn94xQwZnj5iiYO:-O8kVlnBBN5Xne66XwmU:bbfrbk4) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/programs/xdp.md#L5)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

nope, explained above
___
*[`docs/book/start/dropping-packets.md` line 68 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hTxNk4CvHJDhTMwPJ:-O8kVr276FSDmKhhw0jE:b-896fix) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/dropping-packets.md#L68)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

let's avoid manually numbering?
</blockquote></details>

Done.
___
*[`docs/book/start/hello-xdp.md` line 5 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSvHE7GYvpiN5vqZf:-O8kVrdvBwC5vgGg665K:bbfrbk4) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/hello-xdp.md#L5)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

relative?
</blockquote></details>

nope, explained above
___
*[`docs/book/start/hello-xdp.md` line 131 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hU2Go8tTPxWxkHrtr:-O8kVsP80sOKzH7z6p1p:b-896fix) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/start/hello-xdp.md#L131)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

let's avoid manually numbering?
</blockquote></details>

Done.
___
*[`examples/kprobetcp/README.md` line 10 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hU7FI8gpGbsR3fc3v:-O8kVtAzB_W-hWnGaw_r:b-896fix) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/examples/kprobetcp/README.md#L10)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

let's avoid manually numbering?
</blockquote></details>

Done.

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 8 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: 17 of 18 files reviewed, 3 unresolved discussions (waiting on @vadorovsky)

___
*[`docs/index.md` line 64 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSx3FBCNxzd42jo1v:-O8lHB1g9qL7ZuHiXLrv:b-h8do49) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/index.md#L64)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

I can't make this exact link relative in the book (see the comment below).

What I could do if you really want to have this link relative, would be including the Kubernetes logo in `docs/assets` (in a git repo, not as a Github asset, which I don't even know how it's being created :roll_eyes: ), so they end up in the mkdocs build. But I'm not sure if it's the right thing to do. Kubernetes logo is not our asset.

I would rather just link to https://raw.githubusercontent.com/cncf/artwork/refs/heads/main/projects/kubernetes/horizontal/color/kubernetes-horizontal-color.svg if you don't mind.
</blockquote></details>

Yep, seems better. Did you want to remove the asset?
___
*[`docs/book/aya/crosscompile.md` line 12 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSZQGBtYmbI4XO9LF:-O8lIfQrAqLZ67RYuFa5:b-jsug77) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/book/aya/crosscompile.md#L12)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

I swear that having all points as `1.` was triggering a markdownlint error when I started this PR, but I can't reproduce it anymore. Perhaps they relaxed that requirement in markdownlint-cli 0.42.0.

Also, in case of this file, there are points which are followed by code blocks. These code blocks are surrounded by blank lines, which breaks auto numbering. When blank lines are removed, auto numbering works, but `MD031/blanks-around-fences` gets triggered.

LMK what's your opinion about this, but I'm leaning towards removing the blank lines and disabling `MD031/blanks-around-fences` check in this part.
</blockquote></details>

That's weird, I just tried this in github and this renders correctly *with* blank lines around fences. Is this a bug in our mkdocs?

1. hello:
    
   ```
   echo hello
   ```

1. goodbye
___
*[`examples/kprobetcp/README.md` line 10 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hU7FI8gpGbsR3fc3v:-O8lJKjm-G5WF1sOYzfp:b-6fojko) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/examples/kprobetcp/README.md#L10)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

this one not done.

**Review by tamird — COMMENTED** [review]
> Reviewed 1 of 9 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: all files reviewed, 1 unresolved discussion (waiting on @vadorovsky)

**Review by tamird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: 17 of 18 files reviewed, all discussions resolved (waiting on @vadorovsky)

___
*[`docs/index.md` line 64 at r1](https://reviewable.io/reviews/aya-rs/book/165#-O8hSx3FBCNxzd42jo1v:-OCEFDb2AAy69mslPc5q:b-k5z9bl) ([raw file](https://github.com/aya-rs/book/blob/2c871e2ee8f95644fffd650eee7f00729911046e/docs/index.md#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yep, seems better. Did you want to remove the asset?
</blockquote></details>

Nevermind, it's not even in the repo.

**Review by tamird — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/book/165)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @vadorovsky)

**Pull request merged**